### ISSUE TITLE
new AlgoliaSearch(appID, apiKey, opts, ..) => algoliasearch(params)

### ISSUE THREAD

**Issue by vvo** [algolia/algoliasearch-client-js#40]
## tl;dr;

```js
var client = algoliasearch(applicationID, apiKey, {
  protocol: 'http', // force using https or http, previously named "method"
  hosts: [], // which hosts to use for all other requests but the first one
  tld: '', // which tld to use when computing default hosts
  timeout: 2000,
  jsonp: null // feature-detect/force/disable JSONP (null/true/false)
});

var helper = algoliasearch.helper(client, indexName, opts);
```

## Boring stuff

The current constructor code has a lot of code going on and can be instanciated with a lot of different options/arity.

Also as it is a constructor, it must be used with `new AlgoliaSearch()`.

I propose to have only one way to instanciate the client, use only named parameters and a factory function instead of a constructor.

This reduces the overhead while using our library.

Current signature is `function(applicationID, apiKey, methodOrOptions, resolveDNS, hosts)` and must be used with new AlgoliaSearch.

New signature will have only named parameters and will be used with `algoliasearch(params)`.

## TODO

- [ ] remove the need of new AlgoliaSearch()
- [ ] remove the need of new AlgoliaSearchHelper()
- [ ] use only named parameters
- [ ] drop `resolveDNS`. It's not used

## References

- named parameters: [hall of api shame: boolean trap](http://ariya.ofilabs.com/2011/08/hall-of-api-shame-boolean-trap.html)
- a very good JS api design: [browserify](https://github.com/substack/node-browserify#var-b--browserifyfiles-or-opts) (in case you did not notice, we love browserify)

**Comment by redox** [comment]
I don't think we want a `dsnHost` option, I would prefer a boolean here. If you want to specify a custom hostname you can use the hosts array; isn't it?

**Comment by vvo** [comment]
I agree we can ditch dsnHost and only provide the hosts options (where there's no shuffle, nor tld appending).

But what about the boolean? Do you want a boolean to deactivate DSN when using the default hosts?

**Comment by redox** [comment]
You're right; I think we can just get rid of that boolean; I use it in the Algolia Indices explorer (because I use the JS client to perform indexing operations as well) but that's only because it's prepended even if I specify the hosts.

**Comment by vvo** [comment]
Oh yes, no need for a boolean since as soon as we give hosts: [] custom hosts, we can do any complex case (no dsn, dsn,. ).

allright
